index.d.ts

```diff
@@ -0,0 +1,149 @@
+    tmpDirPrefix: string;
+    partSize: number;
+}
+
+declare class File{
+    id: string;
+    upload_length: any;
+    upload_defer_length: any;
+    upload_metadata: any;
+    constructor(file_id: string, upload_length: any, upload_defer_length: any, upload_metadata: any);
+}
+
+/**
+ * Based store for all DataStore classes.
+ */
+export declare class DataStore extends EventEmitter{
+    constructor(options : DataStoreOptions);
+    get extensions() : any;
+    set extensions(extensions_array : any);
+    create(req: http.IncomingMessage): Promise<any>;
+    write(req: http.IncomingMessage, file_id?: string, offset?: number): Promise<any>;
+    getOffset(id: string): Promise<any>;
+}
+
+/**
+ * file store in local storage
+ */
+export declare class FileStore extends DataStore{
+    constructor(options : FileStoreOptions);
+    _checkOrCreateDirectory(): any;
+    create(req: http.IncomingMessage): Promise<any>;
+    write(req: http.IncomingMessage, file_id?: string, offset?: number): Promise<any>;
+    getOffset(file_id: string): Promise<any>;
+}
+
+/**
+ * file store in Google Cloud
+ */
+export declare class GCSDataStore extends DataStore{
+    constructor(options: GCStoreOptions);
+    _getBucket(): any;
+    create(req: http.IncomingMessage): Promise<any>;
+    write(req: http.IncomingMessage, file_id?: string, offset?: number): Promise<any>;
+    getOffset(file_id: string): Promise<any>;
+}
+
+/**
+ * file store in AWS S3
+ */
+export declare class S3Store extends DataStore {
+    constructor(options : S3StoreOptions);
+    _bucketExists(): Promise<any>;
+    _initMultipartUpload(file: File): Promise<any>;
+    _saveMetadata(file: File, upload_id: string): Promise<any>;
+    _getMetadata(file_id: string): Promise<any>;
+    _parseMetadataString(metadata_string: string): any;
+    _uploadPart(metadata: any, read_stream: any, current_part_number: number): Promise<string>;
+    _processUpload(metadata: any, req: http.IncomingMessage, current_part_number: number): Promise<number>;
+    _finishMultipartUpload(metadata: any, parts: Array<any>): Promise<string>;
+    _retrieveParts(file_id: string, part_number_marker: string): Promise<Array<any>>;
+    _countParts(file_id: string): Promise<number>;
+    _clearCache(file_id: string): any;
+    create(req: http.IncomingMessage): Promise<any>;
+    write(req: http.IncomingMessage, file_id?: string, offset?: number): Promise<any>;
+    getOffset(file_id: string, with_parts?: boolean): Promise<any>;    
+}
+
+/**
+ * Tus protocol server implements
+ */
+export declare class Server extends EventEmitter{
+    constructor();
+    get datastore() : DataStore;
+    set datastore(store: DataStore);
+    get(path: string, callback: Function): any;
+    handle(req: http.IncomingMessage, res: http.ServerResponse): http.ServerResponse;
+    listen(): http.Server;
+}
+
+export declare const ERRORS: {
+    MISSING_OFFSET: {
+        status_code: 403,
+        body: 'Upload-Offset header required\n',
```

Comment:
Why are the constant values in here? If I understand it correctly, those values are duplicated from the actual JS code.

Reply:
I'm updating code now.
And I have a question somthings.

I understand that developers use constants related to events.
So I need to add constant about event at ts module.
Because when developer use ts module, ts compiler only find the constant at ts module.
TS compiler doesn't find js module.(I'm talking about when developer use "import" syntax)

My question is does developer who using our package use ERRORS constant?
If it's not, I'll remove the ERROR constant.

Reply:
I understand that you must declare the constants in the .d.ts-file, however, I think it should be possible to keep the constant's value outside of .d.ts. So that the constant's value is still only present in the .js file and the .d.ts file just says "hey, the ERRORS constant has this data type".

This is not only about ERRORS but also other constants like EVENT_ENDPOINT_CREATED which have their value inside .d.ts

Reply:
I remove constant's value and leave only declare.